wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileConverter.java

```diff
@@ -0,0 +1,213 @@
+import org.eclipse.che.api.workspace.server.model.impl.SourceStorageImpl;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceConfigImpl;
+
+/** Helps to convert devfile into workspace config and back. */
+public class DevfileConverter {
+
+  public Devfile workspaceToDevFile(WorkspaceConfigImpl wsConfig) {
+
+    if (!isNullOrEmpty(wsConfig.getDefaultEnv()) || !wsConfig.getEnvironments().isEmpty()) {
+      //  throw new UnsupportedOperationException("Given workspace cannot be converted to devfile
+      // since it is contains environments.");
+    }
+
+    Devfile devFile =
+        new Devfile().withSpecVersion(CURRENT_SPEC_VERSION).withName(wsConfig.getName());
+
+    // Manage projects
+    List<Project> projects = new ArrayList<>();
+    wsConfig
+        .getProjects()
+        .forEach(projectConfig -> projects.add(projectConfigToDevProject(projectConfig)));
+    devFile.setProjects(projects);
+
+    // Manage commands
+    Map<String, String> toolsIdToName = parseTools(wsConfig);
+    List<Command> commands = new ArrayList<>();
+    wsConfig
+        .getCommands()
+        .forEach(command -> commands.add(commandImplToDevCommand(command, toolsIdToName)));
+    devFile.setCommands(commands);
+
+    // Manage tools
+    List<Tool> tools = new ArrayList<>();
+    for (Map.Entry<String, String> entry : wsConfig.getAttributes().entrySet()) {
+      if (entry.getKey().equals(EDITOR_WORKSPACE_ATTRIBUTE_NAME)) {
+        String editorId = entry.getValue();
+        Tool editorTool =
+            new Tool()
+                .withType("cheEditor")
+                .withId(editorId)
+                .withName(toolsIdToName.getOrDefault(editorId, editorId));
+        tools.add(editorTool);
+      } else if (entry.getKey().equals(PLUGINS_WORKSPACE_ATTRIBUTE_NAME)) {
+        for (String pluginId : entry.getValue().split(",")) {
+          Tool pluginTool =
+              new Tool()
+                  .withId(pluginId)
+                  .withType("chePlugin")
+                  .withName(toolsIdToName.getOrDefault(pluginId, pluginId));
+          tools.add(pluginTool);
+        }
+      }
+    }
+    devFile.setTools(tools);
+    return devFile;
+  }
+
+  public WorkspaceConfigImpl devFileToWorkspaceConfig(Devfile devFile)
+      throws DevfileFormatException {
+    validateCurrentVersion(devFile);
+    WorkspaceConfigImpl config = new WorkspaceConfigImpl();
+
+    config.setName(devFile.getName());
+
+    // Manage projects
+    List<ProjectConfigImpl> projects = new ArrayList<>();
+    devFile.getProjects().forEach(project -> projects.add(devProjectToProjectConfig(project)));
+    config.setProjects(projects);
+
+    // Manage tools
+    Map<String, String> attributes = new HashMap<>();
+    StringJoiner pluginsStringJoiner = new StringJoiner(",");
+    StringJoiner toolIdToNameMappingStringJoiner = new StringJoiner(",");
+    for (Tool tool : devFile.getTools()) {
+      if (tool.getType().equals("cheEditor")) {
```

Comment:
Should not an exception be thrown if a tool has an unknown type, for now, any not equals `cheEditor` nor `chePlugin`?

Reply:
+